api/eds.proto

```diff
@@ -107,6 +99,17 @@ message LocalityLbEndpoints {
   // If unspecified, each locality is presumed to have equal weight in a
   // cluster.
   google.protobuf.UInt32Value load_balancing_weight = 3;
+
+  // Optional: the priority for this LocalityLbEndpoints.  If unspecified this will
+  // default to the highest priority (0).
+  //
+  // Under usual circumstances, Envoy will only select endpoints for the highest
+  // priority (0).  In the event all endpoints for a particular priority are
+  // unavailable/unhealthy, Envoy will fail over to selecting endpoints for the
+  // next highest priority group.
+  //
+  // Priorities should range from 0 (highest) to N (lowest) without skipping.
+  google.protobuf.UInt32Value priority = 5;
```

Comment:
Above we state that LB weight must sum to 100. How does this interact with priorities? E.g. what if I have a P0 for locality A with LB weight 50, P0 for locality B with LB weight 50, P1 for locality A with LB weight 75 and P1 for locality B with LB weight 25. What happens if locality A loses all its P0 endpoints? We no longer have a simple sum to 100 of the weights..

Reply:
I don't think that's what it's saying, thought that'd be nice.

I think it's saying you add them up and divide. So you could say 50/50 for an even split but you could also less intuitively say 21/21  

So you for a given priority group you can do all weights or no weights, but if you mix and match it's nonsensical.

Anyway, updated a bit for clarity, please suggest further improvements.   I think this would all make more sense if we had a tree structure int he config but I suspect I'm preaching to the choir :-P

Reply:
Why do we limit the max weight to 100 then? We could set it arbitrarily high. Maybe we should do that.

Reply:
Works for me.  @mattklein123 any thoughts?

Reply:
oh!  actually given the way Envoy does weighted hashing I think having an arbitrary limit makes sense.
Current weighted round robin picks evenly then reuses a backend N times in a row where N is the weight.  Until we have that fixed I lean towards (even arbitrary) limits.     We could replace it with a warning comment instead.

Reply:
The current implementation of weighting in the load balancer is (obviously) incredibly stupid and was implemented to solve a specific problem quickly. Most HW LBs do allow weight specifying as arbitrary numbers, and then do the division. This is why this was done originally as it is more "standard" IMO. 

Sorry I'm not fully tracking the convo above. What is the specific decision we are trying to make?

Reply:
General documentation about weights and priorities and how they interact :-)

I think currently weights are arbitrary (it's nice if the weights for a cluster add up to 100 but not expected or enforced) so when I changed language around how weights interact with priorities I made that more clear.  Harvey says if they're arbitrary, why do we limit them to 100?  I say "for now, implementation details and we can remove later".

We could also actually document that weights for a cluster should add up to 100 (and leave the cap of 100 in there) or remove the limit (and hope people don't run afoul of the WRR code)

Reply:
If we make it arbitrary but need a limit, maybe pick a round number like 128, so we don't have folks confuse it with percentage.

Reply:
Why do we need a limit? I think there are two cases of weighting here:
1) Per host weight inside the LB
2) Cluster weights in WRR

IIRC for (1) we do the "add everything up and divide by N" approach
IIRC for (2) we force add up to 100

Is the concern that each approach is different?